openlibrary/plugins/importapi/code.py

```diff
@@ -97,6 +99,23 @@ def parse_data(data: bytes) -> tuple[dict | None, str | None]:
             raise DataError('unrecognized-XML-format')
     elif data.startswith(b'{') and data.endswith(b'}'):
         obj = json.loads(data)
+
+        # Only look to the import_item table if a record is incomplete.
+        # This is the minimum to achieve a complete record. See:
+        # https://github.com/internetarchive/openlibrary/issues/9440
+        # import_validator().validate() requires more fields.
+        required_fields = ["title", "authors", "publish_date"]
+        has_all_required_fields = all(obj.get(field) for field in required_fields)
```

Comment:
I'm going to double check the data format for empty dates -- we want to check that the date is a valid usable date rather than just there is data present. Strings like "None", "N/A", "unknown", "?" etc shouldn't count, but there are a range of date formats, I can't remember which are likely to be encountered.

We're specifically trying to deal with the data from BWB imports, so I'll check the range of possible values in that data.

Reply:
@scottbarnes , if these imports are coming from the promise items like
https://archive.org/details/bwb_daily_pallets_2024-07-30

specifically the `.ProductJSON.PublicationDate` of each entry in the json file, then I see values like this:


```
"20010112"
"20100101"
"20020201"
"20050131"
null
"197704"
"20020331"
"19680101"
"20170910"
"20230914"
"20240107"
"19710101"
"20121226"
"198601"
"197703"
"20041001"
"19881201"
"20121114"
"19860327"
"200310"
"1986"
null
"20081001"
"20040810"
"20170502"
null
"199512"
"200802"
"20120321"
"20210909"
"20180406"
"20120920"
"19870501"
"20090101"
"20121020"
"1976"
```

digits of length between 4 and 8 or JSON `null`, it seems like an empty string is also a rare possibility.

I did not see any of the wilder string values I was worried about. If you can confirm the data source is bwb_daily_pallets_2024-07-30 style JSON `ProductJSON.PublicationDate`s we should be good.

Reply:
@hornc, I fear I may have misunderstood the the requirements and flow chart in #9440.

In terms of this PR, I was contemplating that the data source would be _any_ import source, and that if a record was not complete, it should not be imported.

I think what I misunderstood was the desire for every imported record to eventually be complete. I understood this as a goal of eventually making it a threshold requirement that records be complete, after initially making it a requirement that records be differentiable and sane.

On this understanding, I was hoping to go straight to that point, but now think 'eventually' means in theory, a record that has already imported, can eventually be made complete by virtue of including a strong identifier. Is that correct?

As it stands, this PR would apply to records of _any_ source and would only allow import if the record is complete (after having checked BookWorm/the affiliate server for more metadata). 

In terms of the workflow diagram, under "Complete now or later?", if a record is not currently complete, once a metadata look up is done by the strong ID, the idea is to still import it even if no further metadata is found, correct? So an incomplete but differentiable record with only a title and ISBN would be imported still, is that right?

Also, would you like this differentiable and sane requirement to apply to all imports, or only promise items?

Reply:
@scottbarnes A record "can eventually be made complete by virtue of including a strong identifier. Is that correct?"  I guess in theory. When I was trying to see what could be done to complete some real ones, the existing record really was undifferentiatable. By adding an ISBN, you make a call as to what the correct book is. For some, I think deleting the undifferentiated record is just as good as adding a strong identifier. Basically, if you can do the work to locate the strong identifier, you may as well just import it fresh. Associating with an existing undifferentiated record doesn't really add anything.

An incomplete but differentiable record with only a title and ISBN should be imported. It'd be nice to attempt to get more metadata if possible, but that could be at a later stage by a different process. The goal would be to have it happen before a promised item was scanned. I think you're talking about an attempted lookup that fails to find anything, those need to be imported. We'd have a chance to try other lookups, and I think that's what we want to collect some stats on -- how many of these do we consistently fail to get full metadata for.

The differentiable and sane requirement should apply to all imports. I think undiferentiatable imports from any source could lead to the problem kinds of mixed metadata ending up at scribe. Promise items seem to be the main source, but I think it applies equally to all sources.

Reply:
@hornc, you are correct that with respect to a record with only a title and ISBN, I was talking about _after_ an attempted lookup that failed to find more metadata. 

I added a new commit that is I hope responsive to your vision. Now the following should be true:
- if a record is incomplete, try to make it complete by fetching metadata from BookWorm/the affiliate-server;
- if a record is still incomplete but has a title and a strong identifier, import it;
- if a record is still incomplete and does not have a strong identifier, do not import it.

This means that a look up for additional metadata will occur before the item is scanned (assuming that promise item imports aren't broken for some other reason).

Here, strong identifiers are limited to `isbn_10`, `isbn_13`, and `lccn`. The issue had mentioned OCN, but I am not actually sure what that is--is it an `ocaid`?

Currently this will not delete an undifferentiated record if there is a match, but will instead attempt to add metadata to empty fields on that existing, undifferentiated record. If the desire is still to delete such records, perhaps that could be a separate issue?

Reply:
@scottbarnes those changes sound good, thank you!

An OCN is a OCLC number, which is assigned by OCLC who operate Worldcat https://search.worldcat.org/

Not deleting records is fine for this piece of work.

As for validating that the data fields are sane, we can look at what is required for dates in more detail.

I'll review all the code against the flow diagram and let you know. It seems good so far.